File: contests/lc/712.MinimumASCIIDeleteSumforTwoStrings/solution.go
package main

func minimumDeleteSum(s1 string, s2 string) int {
	m := len(s1)
	n := len(s2)

	dp := make([]int, n+1)
	for j := n - 1; j >= 0; j-- {
		dp[j] = dp[j+1] + int(s2[j])
	}

	for i := m - 1; i >= 0; i-- {
		prevDiag := dp[n]
		dp[n] = dp[n] + int(s1[i])
		for j := n - 1; j >= 0; j-- {
			temp := dp[j]
			if s1[i] == s2[j] {
				dp[j] = prevDiag
			} else {
				del1 := int(s1[i]) + temp
				del2 := int(s2[j]) + dp[j+1]
				if del1 < del2 {
					dp[j] = del1
				} else {
					dp[j] = del2
				}
			}
			prevDiag = temp
		}
	}

	return dp[0]
}

File: contests/lc/712.MinimumASCIIDeleteSumforTwoStrings/solution.py
class Solution:
    def minimumDeleteSum(self, s1: str, s2: str) -> int:
        m, n = len(s1), len(s2)
        dp = [0] * (n + 1)

        for j in range(n - 1, -1, -1):
            dp[j] = dp[j + 1] + ord(s2[j])

        for i in range(m - 1, -1, -1):
            prev_diag = dp[n]
            dp[n] = dp[n] + ord(s1[i])
            for j in range(n - 1, -1, -1):
                temp = dp[j]
                if s1[i] == s2[j]:
                    dp[j] = prev_diag
                else:
                    delete_s1 = ord(s1[i]) + temp
                    delete_s2 = ord(s2[j]) + dp[j + 1]
                    dp[j] = delete_s1 if delete_s1 < delete_s2 else delete_s2
                prev_diag = temp

        return dp[0]
